Document/0x07b-Testing-Network-Communication.md

```diff
@@ -227,3 +227,87 @@ Ensure that critical operations require at least one additional channel to confi
+- What should be pinned, the certificate, the certificate bytes, public key or the thumb?
+- Lifecycle concerns: since the certificates expire, may be revoked or compromised, processes that handle these situations should be implemented.
+- Platform-specific implementation details are discussed in the respective networking chapters.
+
+#### Static Analysis
+
+If source code is available, the test should check for the following issues:
+
+* Apps performing their own X.509 certificate chain validation. They should instead rely on the existing proven implementations available on the platform SDKs or in trusted, updated third party libraries.
+* Apps implementing certificate pinning by using custom trust/CA stores. These are potentially error prone and may result in performing no certificate validation.
+
+Refer to the platform specific networking chapter for additional details.
+
+#### Dynamic Analysis
+
+Dynamic analysis can be performed by launching a MITM attack using your preferred interception proxy<sup>[1]</sup>. This will allow to monitor the traffic exchanged between client (mobile application) and the backend server. If the Proxy is unable to intercept the HTTP requests and responses, the SSL pinning is correctly implemented.
+
+##### Server certificate validation
+
+Start the analysis by testing the application's behavior while establishing secure connection. The test approach is to gradually relax security of SSL/TLS handshake negotiation and check which security mechanisms are enabled.
+
+1. Using Burp set up a proxy in WiFi settings. Make sure that no certificates are added to trust store (iOS: Settings -> General -> Profiles, Android: depends on the vendor and version of the OS) and that tools like SSL Kill Switch are deactivated. Launch your application and check if you can see the traffic in Burp. Any failures will be reported under the 'Alerts' tab. If you can see the traffic, it means that certificate validation is not performed at all! This effectively means that an active attacker can silently do MiTM against your application. If, however, you can't see any traffic and you have information about the SSL handshake failure, proceed to the next step.
+2. Install the Burp certificate, as explained in the Basic Security Testing section. If the handshake is successful and you can see the traffic in Burp, it means that certificate is validated against device's trust store, but the pinning is not performed. The risk is less significant than in the previous scenario, as two main attack scenarios at this point are misbehaving CAs and phishing attacks, as discussed in Basic Security Testing section.
+3. If executing instructions from previous step doesn't lead to traffic being proxied through Burp, it means that certificate is actually pinned and all security measures are in place. However, you still need to bypass the pinning in order to test the application. Please refer to Basic Security Testing section for more information on this.
```

Comment:
it could be as well that there is a difference in behavior: e.g. the pinning is broken, but traffic is being send through. One might have just informed the server that the pinning is broken by sending additional headers or calling some endpoint it would normally never call.